### ISSUE TITLE
Radically different `update_seq` values from registry.npmjs.com

### ISSUE THREAD

**Issue by kemitchell** [npm/registry-issue-archive#44]
`npm i -g json && curl https://registry.npmjs.com 2>/dev/null | json update_seq` produces markedly different values. Running just today, I've seen values in the 3-million-plus, 2-million-plus, and sub-one-million ranges.

Judging by the corresponding `instance_start_time` values, this is because registry.npmjs.com balances requests across multiple back-end processes that aren't perfectly in sync. A few recent samples:

| instance_start_time | update_seq |
| --- | --- |
| 1471429730053374 | 2,352,383 |
| 1471442890832364 | 2,862,457 |
| 1471444334077742 | 2,862,521 |
| 1471504440058866 | 708,602 (!) |
| 1471514888460026 | 3,059,760 |

An observation and a question:
1. One of the instances seems to be really far behind, with `update_seq` < 1,000,000.
2. Will all back-end processes eventually report the same updates, in the order, with corresponding `.seq` values?

**Comment by ashleygwilliams** [comment]
yup, @kemitchell i ran into this on monday. p sure @mmalecki has a fix for it? also pinging @soldair .

**Comment by mmalecki** [comment]
Hey @kemitchell, thanks for the ping @ashleygwilliams!

In our architecture, `registry-frontdoor` (the thing serving `registry.npmjs.org` traffic) talks to several different CouchDB instances holding the actual data through a load balancer. Each US region has its database load balancer and (currently) 4 CouchDBs. That's the reason for differences in `instance_start_time` and `update_seq` you're seeing.

As we bring up new instances, be it because AWS killed the old ones and let us know that they're 'retired' 30 minutes after the fact, or because we simply decide that they're too old, they start replicating from the write root (a blessed source-of-truth CouchDB; it's a bit of a longer chain, but that's not important here). Thus, new CouchDBs will have frighteningly low `update_seq`, while older instances that went through a lot of document update churn will have higher `update_seq` values.

Our solution to this so far has been to (not explicitly enough) not include `update_seq` in our public API contract. However, there is a single CouchDB host that, in its lifetime, will only see increasing `update_seq` values - that's `skimdb.npmjs.com`. That's what npm Enterprise customers follow in order to clone down the entire registry. There are still two drawbacks, however. First, it does not include scoped modules (be it public or private). There's no way to authenticate with it. Second, when this host goes away (again, AWS probably killed it and let us know half an hour later), the `update_seq`, you guessed it, gets blown away. Ops people get sad, and we have to [tell npm Enterprise customers to restart their replication](http://status.npmjs.org/incidents/669mzfs17fkr).

A potential solution to this problem that is on our roadmap is to have a hot spare for CouchDB that literally copies the `registry.couch` file, ensuring that it serves exactly what the current `skimdb.npmjs.com` host is serving, and thus maintaining the `update_seq` continuity.

Another thing we're working on that's semi-related is `replicate.npmjs.com`. It's like SkimDB, except you can (optionally) authenticate with it just like you would with regular registry - by passing an `Authorization` header with a bearer token (or $ENTITY forbid, basic authentication) and it will send you document changes that you're privy to see. It sources its CouchDB feed from a replication host (part of the longer chain mentioned earlier on), and in my experience that going away is somehow rarer than your usual workhorse CouchDB backing `registry-frontdoor` failing. If I didn't break it some minutes ago, you would be able to `curl https://replicate.npmjs.com/` and get a stable `update_seq`.
(Disclaimer: `replicate.npmjs.com` is still heavily WIP, it might eat your cookies or spill your milk while you're not looking.)

As for your question, it's likely that `registry.npmjs.org` will never report a stable `update_seq`, because it's not a priority for us to do so. `skimdb.npmjs.com` and `replicate.npmjs.com` on the other hand are the endpoints where we care about that, and we'll do our best to have them return an ever-increasing, never-blown-away `update_seq`.

Let me know if you have any further questions!

**Comment by kemitchell** [comment]
@mmalecki Thank you very much for a thorough and informative reply!

Thank you also for mention that replicate.npmjs.com has been up and down. I've seen a few `replicate.npmjs.com` followers drop on TCP errors in the last 24, and figured it was something on my end. Saved me a head-scratching session or two.

I do have a couple further questions:
1. Do I understand correctly that, except after downtime, `update_seq` and change `.seq` numbers from `replicate.npmjs.com` will only increase?
2. Do `.seq` numbers from `skimdb.npmjs.com` and `replicate.npmjs.com` correspond?

**Comment by mmalecki** [comment]
@kemitchell :+1:

The good news is that `replicate.npmjs.com` should be getting more stable now - we deployed a two line fix that drastically reduced memory usage this morning. I believe that was the main reason it kept going down.
1. That's correct. We won't be changing the CouchDB server handling that traffic without a serious reason to do so. If we have to do it, we'll announce that, likely ahead of time.
2. No. `skimdb.npmjs.com` is a special snowflake in that it's populated by a follower that excludes scoped modules. That follower indeed follows the same CouchDB that `replicate.npmjs.com` follows, but effectively "rewrites" it into `skimdb.npmjs.com`, which gets us a whole new set of different `update_seq`'s. Thus, that wouldn't even really be possible - `replicate.npmjs.com`'s `update_seq` increases when a scoped module gets published, `skimdb.npmjs.com`'s doesn't.

**Comment by terinjokes** [comment]
@mmalecki For those of us running our own followers, at what point do you want us to start using `replicate.npmjs.com` in testing? Is subscribing to the status updates the best way to be informed of when that CouchDB server would change (and thus change `update_seq`)?

**Comment by mmalecki** [comment]
Hey @terinjokes, sorry for the radio silence here. Feel free to start testing that endpoint, but it is currently quite unstable due to some performance challenges we ran into. We'll definitely make a more public announcement once we feel it's up to the task!

We'll update status.npmjs.com when we make an `update_seq` change, both on `skimdb.npmjs.com` and `replicate.npmjs.com`.

**Comment by vvo** [comment]
Jumping on here, what is the current accepted flow to be able to: replicate and watch the whole npm registry including scoped public packages?

If I get this well, http://registry.npmjs.com/ can only be used for full single pass replication of all the modules. But since you cannot rely on any update_seq of http://registry.npmjs.com/ you cannot implement today a watch mechanism that would be resilient? cc @zeke any help?

**Comment by bcoe** [comment]
@vvo we should write a blog post on this. We now have the service:

`https://replicate.npmjs.com/registry` which returns global and public scoped packages.

and

`https://skimdb.npmjs.com/registry` which returns just global packages.

I will leave this open as a reminder that we should document this API!

**Comment by vvo** [comment]
Thanks yep right now the information is a bit spread in many places with different outdated/unknown state information. But YAY!

**Comment by zeke** [comment]
@vvo check out https://github.com/npm/registry-follower-tutorial for some clues about how to consume the registry. See also:

- https://github.com/zeke/package-stream uses [changes-stream](https://github.com/jcrugzz/changes-stream) under the hood and can be configured to consume data from replicate.npmjs.com
- https://github.com/zeke/all-the-packages is a streaming offline copy of the skimdb. Does not currently include scoped packages, but easily could.
- https://github.com/zeke/nice-package makes registry data more human-friendly.

**Comment by vvo** [comment]
Thanks! Anyone tried PouchDB? Seems pretty solid for the job

**Comment by zeke** [comment]
https://github.com/nolanlawson/local-npm is an offline registry mirror that uses pouchdb. It's not maintained but you might find some good ideas in its code.

cc @nolanlawson

**Comment by vvo** [comment]
Thanks, I am going the PouchDB way :)
